projects/plugins/jetpack/extensions/shared/use-autosave-and-redirect/index.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * External dependencies
+ */
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useSelect, dispatch } from '@wordpress/data';
+import { useState } from '@wordpress/element';
+
+function redirect( url, callback ) {
+	if ( callback ) {
+		callback( url );
+	}
+	window.top.location.href = url;
+}
+
+export default function useAutosaveAndRedirect( redirectUrl, onRedirect = noop ) {
+	const [ isRedirecting, setIsRedirecting ] = useState( false );
+
+	const { isAutosaveablePost, isDirtyPost, currentPost } = useSelect( select => {
+		const editorSelector = select( 'core/editor' );
+
+		return {
+			isAutosaveablePost: editorSelector.isEditedPostAutosaveable(),
+			isDirtyPost: editorSelector.isEditedPostDirty(),
+			currentPost: editorSelector.getCurrentPost(),
+		};
+	}, [] );
+
+	const isPostEditor = Object.keys( currentPost ).length > 0;
+
+	// Alias. Save post by dispatch.
+	const savePost = dispatch( 'core/editor' ).savePost;
+
+	// For the site editor, save entities
+	const entityRecords = useSelect( select => {
+		return select( 'core' ).__experimentalGetDirtyEntityRecords();
+	} );
+
+	// Save
+	const saveEntities = async () => {
+		for ( let i = 0; i < entityRecords.length; i++ ) {
+			// await is needed here due to the loop.
+			await dispatch( 'core' ).saveEditedEntityRecord(
+				entityRecords[ i ].kind,
+				entityRecords[ i ].name,
+				entityRecords[ i ].key
+			);
+		}
+	};
+
+	const autosave = async event => {
+		event.preventDefault();
+
+		if ( isPostEditor ) {
+			/**
+			 * If there are not unsaved values, return.
+			 * If the post is not auto-savable, return.
+			 */
+			if ( isDirtyPost && isAutosaveablePost ) {
+				await savePost( event );
+			}
+		} else {
+			// Save entities in the site editor.
+			await saveEntities( event );
+		}
+	};
+
+	const autosaveAndRedirect = async event => {
+		event.preventDefault();
+
+		if ( ! window?.top?.location?.href ) {
+			return;
+		}
+
+		// Lock re-redirecting attempts.
+		if ( isRedirecting ) {
+			return;
+		}
+
+		setIsRedirecting( true );
+
+		autosave( event ).then( () => {
+			redirect( redirectUrl, onRedirect );
+		} );
+	};
+
+	return [ autosave, autosaveAndRedirect, isRedirecting ];
```

Comment:
Commented separately in https://github.com/Automattic/jetpack/pull/19671/files#r623519176, but would it make using the hook a bit neater if we returned an object with these methods instead of an array? I _think_ returning an array is more useful when the code consuming the hook intends to rename the methods (like when we use `useState`). For this hook, since we're keeping the method names the same, an object might be more straightforward so in consuming code we can use `const { autosaveAndRedirect } = useAutosaveAndRedirect( redirectUrl )` instead of needing to specify the position of the array to grab that method. (That said, I haven't written many custom hooks, so apologies if there's a reason you'd prefer to use an array!)

Reply:
Not at all -- just a holdover from imitating the structure of the hook this was refactored out of. Returning an object looks much neater for the destructuring, thanks!